### PR TITLE
[R] Upgrade moto to 2.0.2 (#1718)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by amarjandu** [comment]
- [ ] Verify sandbox IT passes
- [ ] Push commit merges to github PR branch

**Comment by amarjandu** [comment]
Removed Jesse from review because hes on vacation 🍹

File: src/azul/service/storage_service.py
Comment (noah-aviel-dove) [reply src/azul/service/storage_service.py:119]: Every invocation of this method uses the `kwargs` to specify the region and never uses it for any other purpose. I can't think of any circumstance where we wouldn't want to be explicit about the region, so I think the `kwargs` should be replaced with a non-optional parameter "`region`" and the construction of the `CreateBucketConfiguration` dictionary should be moved into this method.

File: test/health_check_test_case.py
Comment (noah-aviel-dove) [inline test/health_check_test_case.py:200]: The phrasing strikes me as a little redundant here.
```suggestion
                        'up': False, 'error': f'Queue {queue_name} does not exist for this wsdl version.'
```

File: test/retorts.py
Comment (noah-aviel-dove) [inline test/retorts.py:34]:
```suggestion
    def __init__(self, passthru_url: str = None, request_mock: Optional[responses.RequestsMock] = None) -> None:
```

File: src/azul/service/storage_service.py
Comment (noah-aviel-dove) [reply src/azul/service/storage_service.py:119]: Even better might be to add a `region` attribute to `StorageService`, with the option to override via a parameter like with `bucket_name`, and add a `storage_service` cached property to `AzulUnitTestCase` that automatically sets the region to `self._aws_test_region`. This would however clutter up the superclass with a feature most subclasses wouldn't need, so maybe run this suggestion by Hannes first.

File: test/retorts.py
Comment (noah-aviel-dove) [reply test/retorts.py:34]: (line also probably too long)

File: test/retorts.py
Comment (noah-aviel-dove) [inline test/retorts.py:32]: Seems that some part of the old docstring should be preserved here. Maybe just the last two sentences.

File: requirements.txt
Comment (noah-aviel-dove) [inline requirements.txt:29]: Trailing whitespace

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
PR history is too granular. At the very least, it seems like a mistake to have a commit that updates the `requirements.*` and doesn't include _any_ code changes to reflect the updated libraries. Most if not all of the following commits seem like they should be squashed into that first one, which should get he issue tag. I'm guessing this is an artifact of the development process, not your intent for the merged PR.

**Comment by amarjandu** [comment]
@hannes-ucsc "We can remove the moto backend resets in AzulUnitTestCase, because we cant reproduce the collateral failure in test_portal_service on 2.0.2"

File: test/health_check_test_case.py
Comment (amarjandu) [reply test/health_check_test_case.py:200]: In the HealthController we just pass the message that was raised from the response error.
https://github.com/DataBiosphere/azul/blob/70ca8dbb0f617aac69d73bf12933c4b1dbf3e62d/src/azul/health.py#L111

This was update to match what moto 2.0.2 now returns as a response, I agree that the phrasing is off.

File: test/retorts.py
Comment (amarjandu) [reply test/retorts.py:32]: Check this out, let me know if additional changes should be made.

**Comment by amarjandu** [comment]
Unable to push `417c95c Add StorageServiceTestCase` as an independent commit before the split commit, the introduction of a mixin that uses `@mock_s3` causes issues in moto 1.3.13 w/ the response mocking. Going to require all commits to be within the split commit.

**Comment by amarjandu** [comment]
>@hannes-ucsc "We can remove the moto backend resets in AzulUnitTestCase, because we cant reproduce the collateral failure in TestPortalService on 2.0.2; also remove the bucket cleanup from the TestPortalService.teardown"

* Was not able to remove `moto_backend_reset` from `AzulUnitTestCase.setup` see drop commits
* Was able to remove the bucket teardowns in `TestPortalService` and rely on resets.

**Review by noah-aviel-dove — APPROVED** [review]
The number of split commits is very surprising. The last PR with more than 4 split commits was over three years ago (b300cd93) and was a unique outlier up until this point. However, I can't assert with confidence that this history isn't appropriate given the far-reaching nature of these changes. The changes themselves all look good, so I'll approve, but I encourage you to spend a little more time pondering whether this degree of granularity will be helpful post-merge, i.e., for reverting specific changes while keeping others.

**Comment by amarjandu** [comment]
@noah-aviel-dove 
> The changes themselves all look good, so I'll approve, but I encourage you to spend a little more time pondering whether 
this degree of granularity will be helpful post-merge, i.e., for reverting specific changes while keeping others.

That's the problem with this PR, the granularity helps manage the development/review process because of how wide spread the changes are. Once this is merged into `develop` I don't think we would want to cherry-pick commits to retain while leaving others back. If we want something from this PR we have to take the whole thing. 

An alternative might be to squash the commits down after a final review, before handing it off to the operator, this way at least the review process itself will be easier to manage.

Something like this...
```
* [R 1/4] Update moto to 2.0.2 + Remove workarounds used for 1718
* [R 2/4] Update tests to work with moto 2.0.2
* [R 3/4] Merge ResponseHelper into AuthResponseHelper
* [R 4/4] Move ResponseHelper to attic
```

 This could help keep the develop branch concise, and the usage of the split commit can indicate that there's an everything or nothing commitment to the changes.

**Comment by amarjandu** [comment]
Consider squashing branch down into <5 commits, also `make requirements_update` should be ran before handing it off to an operator.

File: test/health_check_test_case.py
Comment (hannes-ucsc) [inline test/health_check_test_case.py:118]: The commit that touches this line should remove the ResponseHelper and deal with AuthResponseHelper. Or create FIXME+ticket to do that.

File: test/version_table_test_case.py
Comment (hannes-ucsc) [inline test/version_table_test_case.py:33]: I realize that this was inconsistent with the table definition in TF config, but why do we not declare non-key attribute in TF?

cc: @noah-aviel-dove

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:204]: Why was the changed? The use of the dynamically create exception classes superior to poking around in the response.

File: test/retorts.py
Comment (hannes-ucsc) [inline test/retorts.py]: Ignore my earlier comment. But this commit should be squashed with that one. That would have avoided my confusion. And ResponsesHelper should still be deleted.

File: attic/tests/retorts.py
Comment (hannes-ucsc) [inline attic/tests/retorts.py:4]: The whole response helper stuff should be in one commit.

File: test/retorts.py
Comment (hannes-ucsc) [inline test/retorts.py:54]: This looks like it does everything that ResponsesHelper did. If we can eliminate the workarounds that ResponsesHelper implements from many tests why do we have to maintain them here. It looks like your'e also removing the doctest while maintaining the code that the doctest covered. That's a reduction in coverage, if I am not mistaken.

File: test/service/test_portal_service.py
Comment (hannes-ucsc) [inline test/service/test_portal_service.py:109]: useless override

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/version_table_test_case.py
Comment (noah-aviel-dove) [reply test/version_table_test_case.py:33]: I have no idea. I wrote the TF template over a year ago.

File: test/retorts.py
Comment (amarjandu) [reply test/retorts.py]: AuthResponseHelper was removed w/ #2969

File: test/health_check_test_case.py
Comment (amarjandu) [reply test/health_check_test_case.py:118]: ill merge the commit that attics the retorts file with this commit.

File: test/retorts.py
Comment (amarjandu) [reply test/retorts.py:54]: AuthResponseHelper  was removed w/ #2969

File: test/version_table_test_case.py
Comment (amarjandu) [reply test/version_table_test_case.py:33]: it looks like `non_key_attributes` are only available for a `global_secondary_index` https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table

From inspecting the object_versions table for my local deployment it does not look like we use a `global_secondary_index`.

File: src/azul/portal_service.py
Comment (amarjandu) [reply src/azul/portal_service.py:204]: @hannes-ucsc  

I need to fix this; there was a problem with the reproduction that I posted here. https://github.com/databiosphere/azul/issues/1718#issuecomment-801527437

There are two problems here, one was my replication of the invalid version key; i was using an invalid key, but it was also syntactically invalid. Using something like `invalid_version_key` will return a `Invalid version id specified response` not a `NoSuchObjectVersion` exception.

In order to get a NoSuchObjectVersion exception we have to pass a invalid version ID that is syntactically correct. (Something that our tests currently do NOT do https://github.com/DataBiosphere/azul/blob/28e48bb45eae0abaa4d9c4e9d72244cd4ea6e5d2/test/service/test_portal_service.py#L146)

It looks like moto 2.0.2 does NOT inspect for syntactically incorrect version IDs, (they use uuid4s for the version id, rather than the AWS version id)
https://github.com/spulec/moto/blob/f5e3cd891f59a4aa2db3b04fd14cb3e243a1ca52/moto/s3/responses.py#L1210

I need to double check to confirm that this is a bug with moto and raise an issue if my finding match my speculations.

Comment (amarjandu) [reply src/azul/portal_service.py:204]: Replicated here: https://gist.github.com/amarjandu/334ffd668dd849991ba9949a32dcc3b8

Comment (amarjandu) [reply src/azul/portal_service.py:204]: Raised: https://github.com/spulec/moto/issues/3884

Comment (amarjandu) [reply src/azul/portal_service.py:204]: @hannes-ucsc the maintainers team decided that would prefer to keep the moto as feature incomplete (perform no validation) rather than inconsistent (s3api does not validate uuids as versionIDs). https://github.com/spulec/moto/pull/3887 was merged in as a fix to the incorrect exception that was being raised. It should be available as of moto >= 2.0.6.dev14 .

Should we hold off until 2.0.7 is available?

File: test/service/test_portal_service.py
Comment (amarjandu) [inline test/service/test_portal_service.py:138]: Q for @hannes-ucsc

Not sure if this is a required change, however it would more close to an exception from the s3api.

File: src/azul/portal_service.py
Comment (amarjandu) [inline src/azul/portal_service.py:203]: Q for @hannes-ucsc:

This exception handling might not be surgical enough, ClientError casts a wide net of exceptions it will accept, and might hide an actual issue, should we cast wide and filter down? I think `NoSuchKey` would be good to except but that is not the correct exception for an `MissingVersionId`.

https://github.com/spulec/moto/issues/3884

Comment (amarjandu) [reply src/azul/portal_service.py:203]: alternatively we could also remove the Azul class `NoSuchObjectVersion` and just raise the exception that was caught.

File: test/service/test_portal_service.py
Comment (amarjandu) [reply test/service/test_portal_service.py:138]: update to `3sL4kqtJlcpXroDTDmJ+rmSpXd3dIbrHY+MTRCxf3vjVBH40Nr8X8gdRQBpUMLUo`

add link to https://docs.aws.amazon.com/AmazonS3/latest/userguide/versioning-workflows.html

File: src/azul/portal_service.py
Comment (amarjandu) [reply src/azul/portal_service.py:203]: explain hypothesis of eventual-constancy on line 199

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:205]:
```suggestion
            # We hypothesize that when S3 was only eventually consistent, NoSuchKey would have been raised when an object had been
            # created but hadn't materializied yet.
```

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:209]:
```suggestion
            # We hypothesize that when S3 was only eventually consistent, NoSuchVersion would have been raised when the object had
            # been overwritten but the overwrite hadn't materialized yet.
```

Please be more meticulous with your spelling.

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:213]:
```suggestion
                raise
```

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:210]: Why the substring search? The AWS documentation clearly states that the error code *is* `NoSuchVersion`.

File: test/service/__init__.py
Comment (hannes-ucsc) [inline test/service/__init__.py:131]:
```suggestion
    A mixin for test cases that utilize StorageService.
```

File: test/version_table_test_case.py
Comment (hannes-ucsc) [reply test/version_table_test_case.py:33]: > I have no idea. I wrote the TF template over a year ago.

That's a lame response, @noah-aviel-dove. If you wrote this code, you are accountable. I also forget things about code I wrote, but if I do, I dive in and come back with an answer.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by hannes-ucsc** [comment]
@amarjandu 
image

If you squashed old fixups, why are there so many of them on this branch?

**Comment by amarjandu** [comment]
> @amarjandu
> image
> 
> If you squashed old fixups, why are there so many of them on this branch?

@hannes-ucsc my current understanding is this:

If there has been review feedback and changes are required, I need to create fixups for them. Each fixup has to be targeted toward to the correct commit.

This allows a reviewer to look at the comments they left in a previous cycle inspect for comments, 👍🏽 , 👀. If there is a 👍🏽 its should be easy to find a fixup commit that relates to the comment.

During the first cycle that you saw the PR there were 10 splits, in the last cycle there are 8 splits, some that have fixups that you have requested.

On the next cycle, these fixups will be squashed down, and there will be a new set of fixups that you can inspect.

Please let me know if there is something wrong here, I was trying to make it easier to showcase the 👍🏽  and the relative fixups.

File: src/azul/portal_service.py
Comment (amarjandu) [reply src/azul/portal_service.py:210]: I was originally inspecting the exception message (An error occurred (NoSuchVersion) when calling the GetObject operation: The specified version does not exist.) and not the response. I later changed to inspect the e.response, but had left the inspection over.

Changed to string comparison.

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:206]:
```suggestion
            # created but hadn't materialized yet …
```

File: src/azul/portal_service.py
Comment (hannes-ucsc) [inline src/azul/portal_service.py:211]:
```suggestion
            # … and that NoSuchVersion would have been raised when the object had
            # been overwritten but the overwrite hadn't materialized yet.
```

File: test/health_check_test_case.py
Comment (hannes-ucsc) [inline test/health_check_test_case.py:281]: A comment should explain the intent. I usually answer the assumed question "What would happen if this weren't there?"

File: test/service/test_portal_service.py
Comment (hannes-ucsc) [inline test/service/test_portal_service.py:135]:
```suggestion
            # The version identifier below is syntactically correct, but does not refer to
            # any real version of any S3 object. See also https://github.com/spulec/moto/issues/3884
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
A bunch of documentation nits. Rest looks good. I did a full pass.

File: test/health_check_test_case.py
Comment (hannes-ucsc) [inline test/health_check_test_case.py:273]: > Configure the response helper

This part is obvious.

> so a test is enabled to mock out endpoints that might not be used.

This part is incomprehensible.

Take my advice from the previous cycle:

> I usually answer the assumed question "What would happen if this weren't there?"

Comment (hannes-ucsc) [reply test/health_check_test_case.py:273]: After you revise this, consult with someone else on the team to check if they understand what you mean.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
See REVIEW commit. I did test the changes but they may have to be distributed over your existing commits. The change that removes the `@contextmanager` from `_mock_service_endpoints` should probably be a separate commit.

**Comment by hannes-ucsc** [comment]
Alternatively, it's ok if you leave my REVIEW commit in place and just make it the `9/9` commit with a subtitle of like "Refactor response mock usage in health check tests`

**Comment by amarjandu** [comment]
> Alternatively, it's ok if you leave my REVIEW commit in place and just make it the `9/9` commit with a subtitle of like "Refactor response mock usage in health check tests`

I'll update the commit titles and refresh the transitive dependencies.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
I think the passing tests are demonstration enough. I know how hard this was and don't want to strip you of the opportunity to show this off. If you WANT to demo this, feel free to do so and take the `no demo` label off.

**Pull request merged**
